wlib/stl/Map.h

```diff
@@ -0,0 +1,772 @@
+         */
+        const_iterator erase(const_iterator& pos);
+        /**
+         * Erase
+         * @param type
+         * @return
+         */
+        bool erase(key_type& type);
+
+        /**
+         * Returns the value corresponding to a provided key.
+         * @param key the key for which to find the value
+         * @return the value mapped to by the key
+         * @throws KeyException if the key does not map to a value
+         */
+        val_type& at(const key_type& key);
+        /**
+         * @see HashMap::at()
+         * @param key key for which to find the value
+         * @return the mapped value
+         * @throws KeyException if the key does not exist
+         */
+        const val_type& at(const key_type& key) const;
+        /**
+         * @param key key for which to check existence of a value
+         * @return true if the key maps to a value
+         */
+        bool contains(const key_type& key) const;
+        /**
+         * Return an iterator to the map element corresponding
+         * to the provided key, or pass-the-end if the key does
+         * not map to any value in the map.
+         * @param key the key to map
+         * @return an iterator to the element mapped by the key
+         */
+        iterator find(const key_type& key);
+        /**
+         * @see HashMap::find()
+         * @param key the key to map
+         * @return a const iterator to the element mapped by the key
+         */
+        const_iterator find(const key_type& key) const;
+
+        /**
+         * Access an element in the hash map by the given key.
+         * If the key does not map to any value in the map,
+         * then a new value is created and inserted using the default
+         * constructor.
+         * @param key the key whose value to access
+         * @return a reference to the mapped value
+         */
+        val_type& operator[](const key_type& key);
+    };
+
+    void HashMap::init_buckets(HashMap::size_type n)  {
+        m_buckets = static_cast<map_node**>(m_array_allocator.Allocate());
+        for (size_type i = 0; i < n; i++) {
+            m_buckets[i] = nullptr;
+        }
+    }
+
+    void HashMap::ensure_capacity() {
+        if (m_num_elements * 100 < m_max_load * m_max_elements) {
+            return;
+        }
+        size_type new_max = static_cast<size_type>(m_max_elements * 2);
+        Allocator new_array_allocator = Allocator(new_max * sizeof(map_node*), new_max * sizeof(map_node*));
+        map_node** new_buckets = static_cast<map_node**>(new_array_allocator.Allocate());
+        for (size_type i = 0; i < m_max_elements; ++i) {
+            if (!m_buckets[i]) {
+                continue;
+            }
+            map_node* node = m_buckets[i];
+            new_buckets[bucket_index(node->m_key, new_max)] = node;
+        }
+        m_array_allocator.Deallocate(m_buckets);
+        m_buckets = new_buckets;
+        m_array_allocator = new_array_allocator;
```

Comment:
Right here

Reply:
I would suggest using memory_alloc and memory_free instead of Allocator. This is because Allocator is of only one size while Memory uses different sizes. It should but I will have to have a proper look. Will do that by end of today